### PR TITLE
More models (with relationships)

### LINKED ISSUE #11: Add bakery model to API
We need to have a table / model which represents bakeries in our API. Currently, the bakeries information lives in the [pangeo-forge/bakery-database](https://github.com/pangeo-forge/bakery-database) github repo as a single yaml file. A key question going forward is **whether we want to keep the yaml file as our "source of truth" for the bakeries or just make the database itself our source of truth**.

### Option 1: Keep Github yaml file as source of truth

_Pros:_ consistency with existing setup; transparent, familiar way to add / update bakeries

_Cons:_ need to keep database in sync with yaml file; how to deal with inconsistent data in yaml file?; how to trigger updates to database when yaml file is updated

### Option 2: Drop Github yaml file, go all in on database

In this option, we would just directly enter / update the bakery information via the REST API or command line.

_Pros:_ single source of truth; integrates seamlessly with rest of database

_Cons:_ slightly less transparent; need to ensure consistency across migrations

---

My vote is that we should go with option 2 and just drop the github yaml file. Keeping the db in sync with github is a problem we should not waste time on. The cons of option 2 can be mitigated via good documentation and continuous deployment practices.

### LINKED ISSUE #12: Add feedstock model to API
According to our [database schema](https://dbdiagram.io/d/6192974202cf5d186b57003f) we need a table / model for feedstocks. This table may be as simple as a mapping between `feedstock_id` and github repo. Some questions

- How do we store github repo? Some options:
  - `github_repo: "https://github.com/pangeo-forge/my-awesome-feedstock"`
  - `github_repo: "pangeo-forge/my-awesome-feedstock"`
  - `github_org: "pangeo-forge", github_repo: "my-awesome-feedstock"`
- Do we need to keep track of a specific branch (e.g. main, master, etc.)?
- Do we perform any validation of the feedstock when creating a new feedstock entry? Examples of validation
  - github repo exists
  - github repo is a valid feedstock repo (e.g. contains `meta.yaml`)
- How do we handle the special repo `pangeo-forge/staged-recipes`, which will generate many recipe runs as part of the PR workflow?

### PR DESCRIPTION
**Pull request by cisaacstern**
When complete

- closes #11
- closes #12 

These models (along with the others in [our ERD](https://dbdiagram.io/d/6192974202cf5d186b57003f)) will also be linked via [Relationship Attributes](https://sqlmodel.tiangolo.com/tutorial/fastapi/relationships/) (see also [here](https://sqlmodel.tiangolo.com/tutorial/relationship-attributes/define-relationships-attributes/)).

### CODE REVIEW

**Comment by rabernat** [comment]
Great! Hold off on much testing of this as I have a big testing refactor PR coming

File: migrations/versions/22828222f2fe_add_bakery_table.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    op.create_index(op.f("ix_reciperun_completed_at"), "reciperun", ["completed_at"], unique=False)
+    op.create_index(op.f("ix_reciperun_conclusion"), "reciperun", ["conclusion"], unique=False)
+    op.create_index(op.f("ix_reciperun_feedstock_id"), "reciperun", ["feedstock_id"], unique=False)
+    op.create_index(op.f("ix_reciperun_head_sha"), "reciperun", ["head_sha"], unique=False)
+    op.create_index(op.f("ix_reciperun_id"), "reciperun", ["id"], unique=False)
+    op.create_index(op.f("ix_reciperun_message"), "reciperun", ["message"], unique=False)
+    op.create_index(op.f("ix_reciperun_recipe_id"), "reciperun", ["recipe_id"], unique=False)
+    op.create_index(op.f("ix_reciperun_started_at"), "reciperun", ["started_at"], unique=False)
+    op.create_index(op.f("ix_reciperun_status"), "reciperun", ["status"], unique=False)
+    op.create_index(op.f("ix_reciperun_version"), "reciperun", ["version"], unique=False)
```
Comment (rabernat) [inline migrations/versions/22828222f2fe_add_bakery_table.py:43]: I don't understand why all these indexes are being created by this migration.

Comment (cisaacstern) [reply migrations/versions/22828222f2fe_add_bakery_table.py]: This migration was deleted in https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31/commits/e64d7eae3ed3ff5f823359d8429aaa79597782cf, but the underlying question remains a mystery.

We are not defining any [SQLModel indexes](https://sqlmodel.tiangolo.com/tutorial/indexes/), so it's not that.

AFAICT, Alembic uses SQLAlchemy's [`get_indexes`](https://alembic.sqlalchemy.org/en/latest/autogenerate.html#controlling-what-to-be-autogenerated) inspection to generate these `create_index` calls. Specifically, that's called as a [method of an `inspector` object in `alembic.autogenerate.compare`](https://github.com/sqlalchemy/alembic/blob/85152025ddba1dbeb51b467f40eb36b795d2ca37/alembic/autogenerate/compare.py#L558).

So it seems like these indexes must be inferred from some other attributes on that `inspector` object, or perhaps (more likely?) from the `schema` kwarg being passed to `get_indexes`. There's some type of semi-opaque decorator registration happening in Alembic there, so I didn't get to the point of actually tracking down where, e.g., that `schema` value is defined.

**Comment by cisaacstern** [comment]
Ok I see from the failed tests that even though local tests don't require an alembic migration, I do need to generate and commit a new migration version in order for the Heroku CI to pass. I guess I will just do that once the PR is otherwise complete. Otherwise, I would either be **(a)** generating a new version every time I push to remote (which would result in too many versions); or **(b)** overwriting a single migration version each time I push to remote (which would lead a messy commit history, especially because the version module names include a hash).

**Comment by rabernat** [comment]
Correct. I think a new migration needs to be generated every time the table models change. Ideally such changes should be minimal (no back and forth in the commit history), and the migration script should be in the same commit as the model changes.

**Comment by cisaacstern** [comment]
>  and the migration script should be in the same commit as the model changes

Can you elaborate on this? Presumably this so it's clear what model changes the migration relates to? If so, is _same PR_ (rather than same commit) sufficient?

In this PR, I'm of course already deviating from the same commit pattern. It seems like a natural development process to incrementally change the models (one feature at a time) as part of a PR, testing locally along the way, and then generate a migration only when the PR is otherwise ready to merge. But this would spread the model changes over many commits, and put the migration in its own commit afterwards. Open to alternative practices, though.

**Comment by rabernat** [comment]
> If so, is _same PR_ (rather than same commit) sufficient?

Sure.

"Model changes" means any changes / additions / deletions to the `SQLModel(table=true)` classes, which alembic uses to generate migrations.

File: README.md [hunk 2]
```diff
@@ -82,6 +82,9 @@ with tables based on `pangeo_forge_orchestrator.models`. If a file already exist
 to refect the tables defined by `pangeo_forge_orchestrator.models`. Note that `.sqlite` files are excluded by `.gitignore`, so you don't
 need to worry about the database file appearing in your commit.
 
+**Note:** Changes to modules in `pangeo_forge_orchestrator/*` may require that the package be re-installed with `pip install '.[dev]'`
+before running local tests.
+
```
Comment (cisaacstern) [inline README.md:87]: Maybe this is just a basic pytest assumption, in which case we can remove. But I kept forgetting to do this and spent unnecessary amounts of time wondering why the tests were failing. 🤦

Comment (rabernat) [reply README.md:87]: If you run `pip install -e '.[dev]'` once, this should be unnecessary.

**Comment by cisaacstern** [comment]
@rabernat, the Bakery and Feedstock tables and associated relationships are now implemented. I though I would pause for a review here so that the `sqlmodel.Relationship` implementation in `model_builders` and `model_fixtures` doesn't get obscured by any further work (i.e. migrations, etc.) that we'll want to do on this PR.

My hope is that the docstrings and comments (including SQLModel docs links therein) are sufficient to make the implementation clear. But if not, please let me know. I'll make an inline note about [`back_populates`](https://sqlmodel.tiangolo.com/tutorial/relationship-attributes/back-populates/), which I could use some advice on.

If/when this phase of work looks good to you, we can choose to either

- Run alembic migrations and merge, then revisit other work as separate PRs

or do any/all of the following as part of this PR, before running migrations

- Tighten field constraints, e.g. maybe:
    - `Enum` type for `Bakery.region` instead of a plain string
    - [`pydantic.HttpUrl`](https://pydantic-docs.helpmanual.io/usage/types/#urls) for `Feedstock.github_repo`
    - etc.
- Add the remainder of the models from our [ERD](https://dbdiagram.io/d/6192974202cf5d186b57003f): `dataset`, `person`, `bakery_maintainers`, `feedstock_maintainers`

File: pangeo_forge_orchestrator/models.py [hunk 3]
```diff
@@ -1,10 +1,54 @@
...
+# Feedstock -----------------------------------------------------------------------------
+
+
+class FeedstockBase(SQLModel):
+    """Information about a Pangeo Forge Feedstock.
+
+    :param github_repo: The url of the associated GitHub repo.
+    """
+
+    github_repo: str  # TODO: Pydantic URL constraint.
```
Comment (rabernat) [inline pangeo_forge_orchestrator/models.py:41]: Pros and cons of this vs
```suggestion
    github_org: str   # org name, e.g pangeo-forge
    github_repo: str  # just the repo name e.g. staged-recipes
```
?

**Review by rabernat — APPROVED** [review]
This looks good.

Are we going to support _updating a relationship_? Like if a recipe_run is created pointing to a specific bakery, can we update that to reference a different bakery?

If so, we need to test this update.

If not, we need to test that such an update attempt fails.

File: pangeo_forge_orchestrator/models.py [hunk 4]
```diff
@@ -81,6 +125,67 @@ class RecipeRunRead(RecipeRunBase):
...
+
+bakery_models = MultipleModels(
+    path="/bakeries/",
+    base=BakeryBase,
+    response=BakeryRead,
+    extended_response=BakeryReadWithRecipeRuns,
+    relations=[
+        RelationBuilder(
+            field="recipe_runs",
+            back_populates="RecipeRun.bakery",
```
Comment (cisaacstern) [inline pangeo_forge_orchestrator/models.py:156]: @rabernat, note here that I'm passing `back_populates` as a dot-delimited string, which is then [`split()` in `model_builders.MultipleModels.make_table_cls`](https://github.com/pangeo-forge/pangeo-forge-orchestrator/blob/fa0392fd3d29b15b124b174fba779e0fe14c66d3/pangeo_forge_orchestrator/model_builders.py#L128).

`sqlmodel.Relationship`, [for which this is a kwarg value](https://sqlmodel.tiangolo.com/tutorial/relationship-attributes/back-populates/), just wants "the name of the attribute in the other model class" (i.e. "bakery"), not a string inclusive of the name of the other model class (i.e. "RecipeRun.bakery").

But for some reason, I just find this _very_ difficult to read without including the model name in the string. I don't love doing this, because it breaks @tmcw's guidance to ["call things the same thing"](https://macwright.com/2021/02/17/the-naming-of-things.html).

Curious how this reads to you. If we keep it this way, I can clarify what we're looking for in [`RelationBuilder` docstring](https://github.com/pangeo-forge/pangeo-forge-orchestrator/blob/fa0392fd3d29b15b124b174fba779e0fe14c66d3/pangeo_forge_orchestrator/model_builders.py#L17). Open to changing it though.

File: pangeo_forge_orchestrator/model_builders.py [hunk 5]
```diff
@@ -90,12 +115,18 @@ def make_table_cls(self) -> SQLModel:
...
+        https://sqlmodel.tiangolo.com/tutorial/relationship-attributes/define-relationships-attributes/
         """
         cls_name = self.make_cls_name(self.base, "")
         attrs = dict(id=Field(default=None, primary_key=True))
         annotations = dict(id=Union[int, None])
         attrs.update(dict(__annotations__=annotations))
+        if self.relations:
+            for r in self.relations:
+                attrs.update({r.field: Relationship(back_populates=r.back_populates.split(".")[1])})
+                attrs.get("__annotations__").update({r.field: r.annotation})  # type: ignore
```
Comment (rabernat) [inline pangeo_forge_orchestrator/model_builders.py:129]: Can you say in words what this does? How does it transform the `relations` field? Give a before and after example.

File: pangeo_forge_orchestrator/models.py (hunk 4 above)
Comment (rabernat) [reply pangeo_forge_orchestrator/models.py:156]: If the value is being ignored by the code that consumes it, then it should be removed.

Is the `RecipeRun` part of `RecipeRun.bakery` always the same value `annotations=List["RecipeRun"]`. If so, it seems like we could avoid this redundancy.

**Review by rabernat — COMMENTED** [review]
Responses to your questions

File: pangeo_forge_orchestrator/models.py (hunk 3 above)
Comment (cisaacstern) [reply pangeo_forge_orchestrator/models.py:41]: I like the conciseness of passing just a repo name. (Especially assuming we can default `github_org="pangeo-forge"`.)

But in response JSON, it would be nice to have a copy-and-pastable URL? If we go with this suggestion, I guess we could generate the full URL as a @property somewhere.

File: pangeo_forge_orchestrator/models.py (hunk 4 above)
Comment (cisaacstern) [reply pangeo_forge_orchestrator/models.py:156]: > If the value is being ignored by the code that consumes it, then it should be removed.

Yes... it is being ignored. 🤭 

> Is the RecipeRun part of RecipeRun.bakery always the same value annotations=List["RecipeRun"]. If so, it seems like we could avoid this redundancy.

Correct. The `back_populates` field will always be field of the model referenced in `annotations`. This is actually obvious now that I write it out that way, but somehow felt more elusive before. I will fix this to eliminate the redundancy and update the docstring accordingly.

File: pangeo_forge_orchestrator/model_builders.py (hunk 5 above)
Comment (cisaacstern) [reply pangeo_forge_orchestrator/model_builders.py:129]: [Plain language summary]

`.relations` is an optional list of `RelationBuilder`s. If present, it is used in the `.make_table_cls` method to add `sqlmodel.Relationship` attributes to the generated table model, as described in the SQLModel documentation for [declaring relationship attributes](https://sqlmodel.tiangolo.com/tutorial/relationship-attributes/define-relationships-attributes/#declare-relationship-attributes) on a table model. Specifically, each relationship attribute requires a field name (`RelationBuilder.field`), a type annotation (`RelationBuilder.annotation`), and the name of the field in the related model to back-populate (`RelationBuilder.back_populates`). Note that the annotation is either the related table class object (in the case of a one-to-one relationship), or a `List[str]` where `str` is the `__name__` of the related table class object. Either annotation type can be `Optional`.

[Before]

A `RelationBuilder` enters this control flow with the following fields:

- `field`: a name for the field to set on the table model being generated in this method
- `annotation`: a type annotation for the field to be set on the table model being generated in this method
- `back_populates`: the name of the field _in the related model_

[After]

For each of the `RelationBuilder` objects in the list of `self.relations`, the table model model generated in this method has the following fields set on it by the `self.relations` control flow in this method:

```python
from sqlmodel import Relationship, SQLModel

class GeneratedTableModel(SQLModel, table=True):   # the model we're generating in this method

    # ... other fields exist on this model (generated outside this particular `self.relations` control flow)

    RelationBuilder.field: RelationBuilder.annotation = Relationship(back_populates=RelationBuilder.back_populates)

```

> A small caveat: Currently, the `sqlmodel.Relationship.back_populates` kwarg value is not a 1-to-1 with `RelationBuilder.back_populates`, but it _will be_ after the change discussed in https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31#discussion_r789907629 is implemented.

File: pangeo_forge_orchestrator/models.py (hunk 4 above)
Comment (cisaacstern) [reply pangeo_forge_orchestrator/models.py]: https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31/commits/56c0323001c7d3cfb4663486aa0cccd1c47b17fa

**Comment by cisaacstern** [comment]
> Are we going to support updating a relationship? Like if a recipe_run is created pointing to a specific bakery, can we update that to reference a different bakery? ... If not, we need to test that such an update attempt fails.

Pydantic's `Field(allow_mutation=False)` did not work out of the box, so I've split this off into its own issue #35

**Comment by cisaacstern** [comment]
I think the reason the Heroku CI failed is that the alembic migration was generated against a SQLite `DATABASE_URL`, as indicated by the first line of logging it output on the call to autogenerate

```
➜  python -m alembic revision --autogenerate -m "add_bakery_and_feedstock"
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
```

I will try starting a local Postgres server and generating the migration against that.

**Review by rabernat — COMMENTED** [review]
> I think the reason the Heroku CI failed is that the alembic migration was generated against a SQLite `DATABASE_URL`

In my understanding, the migration is not specific to the database. That's kind of the whole point of sqlalchemy / alembic. The migration in #27 worked on both sqlite and postgres. So I think the source of the error must be different. The migration seems to run fine in the [logs](https://dashboard.heroku.com/pipelines/17cc0239-494f-4a68-aa75-3da7c466709c/tests/72):

```
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> 81c7d27bde71, initial migration
INFO  [alembic.runtime.migration] Running upgrade 81c7d27bde71 -> 6197f9c7aff6, add_bakery_and_feedstock
-----> test-setup command `pip install -e '.[dev]' && python -m alembic upgrade head` completed successfully
```

The error instead is this:

```
E       psycopg2.errors.ForeignKeyViolation: insert or update on table "reciperun" violates foreign key constraint "reciperun_bakery_id_fkey"
E       DETAIL:  Key (bakery_id)=(1) is not present in table "bakery".
```

Postgres is not letting you create a reference to a bakery that doesn't exist. That's good! Posgres is enforcing our stated constraints more strictly than Heroku.

This means we will need to modify our testing strategy so that we never create a recipe_run without first creating a valid bakery. A fixture would be perfect for this. That may strain machinery we have created for automatically generating all these tests, but it needs to be done somehow.

File: migrations/versions/22828222f2fe_add_bakery_table.py (hunk 1 above)
Comment (rabernat) [reply migrations/versions/22828222f2fe_add_bakery_table.py]: This looks like https://github.com/tiangolo/sqlmodel/issues/9

**Comment by cisaacstern** [comment]
> Postgres is not letting you create a reference to a bakery that doesn't exist. That's good! Posgres is enforcing our stated constraints more strictly than sqlite.

Thanks for clarifying. Noting for my future self (and to potentially add to the docs later), that this same Postgres error is replicable locally by

- Installing https://postgresapp.com and running `psql` to enter an interactive session
- Running `CREATE DATABASE test_db;` within the interactive session, then exiting
- Setting `export DATABASE_URL=postgresql://localhost/test_db`
- Running alembic migration against the new Postgres `DATABASE_URL`
- `pytest -vx`

We might expect `pytest -vx -k "read_single"` to pass already, because this test is already set up to create the `ModelFixture.requires_relations` tables as a first step. This test does get past the `ForeignKeyViolation` error that we're seeing in CI, but it surfaces what appears to be another Postgres-specific issue:

> E               AssertionError: assert '2020-12-31T16:00:00Z' == '2021-01-01T00:00:00Z'
E                 - 2021-01-01T00:00:00Z
E                 + 2020-12-31T16:00:00Z

which appears to result from the fact that [Postgres time stamps](https://www.postgresql.org/docs/9.2/datatype-datetime.html) are read back in the system's timezone. (Indeed, the time delta reflected in the AssertionError above is the difference between UTC and my computer's timezone.)

I think the easiest way to fix this is to manually set our SQLAlchemy session fixture timezone to UTC, if possible. Once I get this resolved, I'll move on to

> This means we will need to modify our testing strategy so that we never create a recipe_run without first creating a valid bakery. A fixture would be perfect for this. That may strain machinery we have created for automatically generating all these tests, but it needs to be done somehow.

as proposed above.

**Comment by cisaacstern** [comment]
> Postgres time stamps are read back in the system's timezone. 

https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31/commits/0c84a2080b0bbc19185d6ee47074b9ecf53ff047 solves this based on https://stackoverflow.com/a/26106482

> this same Postgres error is replicable locally

Another SQLite vs. Postgres issue! (It's increasingly looking like having the ability to test Postgres locally will be quite helpful..)

Our prior `clear_table` function ensures that each table in the database is empty prior executing a given test. In SQLite, the primary key sequence for an empty table appears to always begin at 1 (regardless of whether rows have been previously created and deleted in the table). In Postgres, the primary keys for deleted rows are apparently _not_ reused by default, so rows added to an empty table will not be assigned a given primary key if that key has previously been used for another row in that table (even if that row was subsequently deleted).

https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31/commits/31565bc075297390df78e3ba05cd7a3d20f43795 fixes this based on https://stackoverflow.com/a/36248764

With the addition of these two commits, I can get tests to pass against a local Postgres database, so will now move onto the aforementioned 

> This means we will need to modify our testing strategy so that we never create a recipe_run without first creating a valid bakery. A fixture would be perfect for this. That may strain machinery we have created for automatically generating all these tests, but it needs to be done somehow.

**Comment by cisaacstern** [comment]
@rabernat, referring back to our conversation in https://github.com/pangeo-forge/pangeo-forge-orchestrator/issues/14#issuecomment-1008984711

> As I/we learn more about the differences (or lack thereof) between SQLite and Postgres, we may decide it's not necessary [to test against both databases].

it seems clear to me now that testing locally against Postgres (as I've now described in the README) is an essential final check to debug differences with SQLite that may make the Heroku CI fail. In working on a solution for the `ForeignKeyError` we saw on Heroku, local Postgres testing ended up surfacing the timezone and primary key sequence issues discussed in https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31#issuecomment-1020601666.

In terms of the `ForeignKeyError` solution, I haven't used a true fixture (as, IIUC, your earlier comment suggested), but rather  combined some new data structures in `model_fixtures.py` with logic in `test_api.py` that looks for them and triages each table's dependencies as needed. This could probably be a bit simpler, and I'm totally open to edits here, but just wanted to push something that worked as a starting place.

File: pangeo_forge_orchestrator/models.py [hunk 6]
```diff
@@ -1,10 +1,63 @@
...
+class FeedstockBase(SQLModel):
+    """Information about a Pangeo Forge Feedstock.
+
+    :param github_repo: The name of the feedstock repo, e.g. ``"noaa-oisst-avhrr-only-feedstock"``.
+    :param github_org: The name of the GitHub org in which this feedstock repo resides. Must be one
+      of the options defined in ``GitHubOrgs``.
+    """
+
+    github_repo: str
+    github_org: GitHubOrgs = GitHubOrgs.pangeo_forge
```
Comment (rabernat) [inline pangeo_forge_orchestrator/models.py:50]: Sorry to flip flop on this so much. But I've really been thinking about the best way to refer to feedstocks. The way this is currently done (based on my prior comment) explicitly assumes github is the provider. It might be nice to future-proof ourselves around the idea of multiple possible providers (e.g. we may one day want to integrate with gitlab).

A great template for how this is done is binder:
- Documentation of different RepoProviders: https://binderhub.readthedocs.io/en/latest/reference/repoproviders.html#zenodoprovider
- Code: https://github.com/jupyterhub/binderhub/blob/master/binderhub/repoproviders.py
- Resolver: https://github.com/jupyterhub/binderhub/blob/c08d5dbadccd5fbd831b8574570bec3219fb4d15/binderhub/main.py#L15-L24

So bottom line, I would recommend changing this to

```suggestion
    provider: RepoProvider = RepoProvider.github
    spec: str
```

where spec is dependent on the provider. For github, it would be a string like `pangeo-forge/my-great-recipe`. (I would not assume that all repos will live in the pangeo-forge org).

File: tests/model_fixtures.py [hunk 7]
```diff
@@ -1,27 +1,43 @@
...
     path: str
     required_fields: Sequence[str]
     create_opts: Sequence[APIOpts]  # valid ways to create the model
     invalid_opts: Sequence[APIOpts]  # setting these on creation or update should error
     update_opts: Sequence[APIOpts]  # setting these on update should be valid
 
 
+@dataclass
+class ModelRelationFixture:
+    field_name: str  # Name of the field cooresponding to the related table
```
Comment (rabernat) [inline tests/model_fixtures.py:21]:
```suggestion
    field_name: str  # Name of the field corresponding to the related table
```

I would also like a one-line comment about what this class does.

File: tests/model_fixtures.py [hunk 8]
```diff
@@ -1,27 +1,43 @@
...
+    model_fixture: ModelFixture
+
+
+@dataclass
+class ModelFixtureWithDependencies(ModelFixture):
```
Comment (rabernat) [inline tests/model_fixtures.py:26]: Could we simplify things by just attaching this attribute to ModelFixture? An empty list (default) would mean no dependencies.

File: tests/model_fixtures.py [hunk 9]
```diff
@@ -1,27 +1,43 @@
...
+    dependencies: List[ModelRelationFixture] = field(default_factory=list)  # req'd to create model
+
+
+@dataclass
+class ModelFixtureWithOptionalRelations(ModelFixture):
+    optional_relations: List[ModelRelationFixture] = field(default_factory=list)  # not req'd
```
Comment (rabernat) [inline tests/model_fixtures.py:32]: same

File: tests/test_api.py [hunk 10]
```diff
@@ -37,14 +37,26 @@ def compare_response(response_fixture, reponse_data):
             assert actual == expected
 
 
-create_params = [
-    (mf.path, create_opts) for mf in ALL_MODEL_FIXTURES for create_opts in mf.create_opts
-]
+def create_with_dependencies(create_opts, mf, client):
+
+    if hasattr(mf, "dependencies"):
```
Comment (rabernat) [inline tests/test_api.py:42]: My comment above would make this `if` unnecessary.

File: tests/model_fixtures.py (hunk 8 above)
Comment (cisaacstern) [reply tests/model_fixtures.py:26]: Oh yeah! I'd initially dismissed this as impossible because of the circular type hinting. The missing link here was that I didn't realize (until just now) that `List["SomeClass"]` ForwardRef-style type hinting could make this possible. Change coming momentarily.

Comment (cisaacstern) [reply tests/model_fixtures.py]: ✔️ https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31/commits/ace3c1a02743c1e740d01906a4bd48a32099d87a

File: tests/model_fixtures.py (hunk 9 above)
Comment (cisaacstern) [reply tests/model_fixtures.py]: ✔️ https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31/commits/ace3c1a02743c1e740d01906a4bd48a32099d87a

File: tests/test_api.py (hunk 10 above)
Comment (cisaacstern) [reply tests/test_api.py]: I think I got all of these out between https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31/commits/ace3c1a02743c1e740d01906a4bd48a32099d87a and https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31/commits/930857c8c4f16d6d14cf42afffce02fdde11b052. Much nicer this way!

File: tests/model_fixtures.py (hunk 7 above)
Comment (cisaacstern) [reply tests/model_fixtures.py]: >  a one-line comment about what this class does.

Ended up being two lines; hopefully this makes sense https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/31/commits/45c5004fb47fe678898fa47eb47f653e263668c8

**Comment by cisaacstern** [comment]
(Heroku CI was stalled on _Test Setup_ 🤷 . Restarted from the Heroku dashboard and we're now passing.)

**Review by rabernat — APPROVED** [review]
I checked this out and tested it locally. Seems to be working great! I think we are done here!

One point of developer friction I experienced was that I needed to manually remove my `database.sqlite` in order to get the new tables to create properly on startup. No action really needed here, but perhaps we can update the docs to clarify this.

**Comment by rabernat** [comment]
I'm not worried about pre-commit. That failure is a tiny artifact of the edit I made on gh.
